File: lib/php/tests/PeersTest.php
<?php

declare(strict_types=1);

namespace WgPortal\Client\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use WgPortal\Client\HttpClient;
use WgPortal\Client\PeersApi;
use WgPortal\Client\WgPortalException;

class PeersTest extends TestCase
{
    private const BASE_URL = 'http://wg-portal.test:8888';
    private const PEER_ID = 'xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=';

    private function makeApi(array &$queue): PeersApi
    {
        $ref = new \ReflectionClass(HttpClient::class);
        $http = $ref->newInstanceWithoutConstructor();

        $prop = $ref->getProperty('baseUrl');
        $prop->setAccessible(true);
        $prop->setValue($http, self::BASE_URL . '/api/v1');

        $headersProp = $ref->getProperty('headers');
        $headersProp->setAccessible(true);
        $headersProp->setValue($http, ['Authorization' => 'Basic dGVzdDp0ZXN0']);

        $mock = new MockHandler($queue);
        $httpProp = $ref->getProperty('http');
        $httpProp->setAccessible(true);
        $httpProp->setValue($http, new GuzzleClient(['handler' => HandlerStack::create($mock)]));

        return new PeersApi($http);
    }

    private function fixturePeer(): array
    {
        return [
            'Identifier' => self::PEER_ID,
            'InterfaceIdentifier' => 'wg0',
            'PrivateKey' => 'k',
            'PublicKey' => 'k',
            'DisplayName' => 'My Peer',
            'UserIdentifier' => 'uid-1234567',
        ];
    }

    public function testGetById(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode($this->fixturePeer()));

        $result = $api->getById(self::PEER_ID);
        $this->assertEquals(self::PEER_ID, $result['Identifier']);
        $this->assertEquals('wg0', $result['InterfaceIdentifier']);
    }

    public function testGetByInterface(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([$this->fixturePeer()]));

        $result = $api->getByInterface('wg0');
        $this->assertCount(1, $result);
    }

    public function testGetByUser(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode([$this->fixturePeer()]));

        $result = $api->getByUser('uid-1234567');
        $this->assertCount(1, $result);
    }

    public function testCreate(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(200, ['Content-Type' => 'application/json'], json_encode($this->fixturePeer()));

        $result = $api->create($this->fixturePeer());
        $this->assertEquals(self::PEER_ID, $result['Identifier']);
    }

    public function testDelete(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(204);

        $api->delete(self::PEER_ID);
        $this->assertTrue(true);
    }

    public function testGetByIdThrowsOn404(): void
    {
        $queue = [];
        $api = $this->makeApi($queue);
        $queue[] = new Response(404, ['Content-Type' => 'application/json'], '{"Code":404,"Message":"Not Found"}');

        $this->expectException(WgPortalException::class);
        $api->getById('nonexistent');
    }
}

File: lib/rust/tests/peers_test.rs
use mockito::Server;
use serde_json::json;
use wg_portal_client::{WgPeer, WgPortalClient};

fn setup() -> (mockito::ServerGuard, WgPortalClient) {
    let server = Server::new();
    let client = WgPortalClient::new(&server.url(), "admin@wgportal.local", "test-token");
    (server, client)
}

const PEER_ID: &str = "xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=";

fn fixture_peer() -> serde_json::Value {
    json!({
        "Identifier": PEER_ID,
        "InterfaceIdentifier": "wg0",
        "PrivateKey": "yAnz5TF+lXXJte14tji3zlMNq+hd2rYUIgJBgB3fBmk=",
        "PublicKey": "TrMvSoP4jYQlY6RIzBgbssQqY3vxI2Pi+y71lOWWXX0=",
        "DisplayName": "My Peer",
        "UserIdentifier": "uid-1234567"
    })
}

fn fixture_peer_list() -> serde_json::Value {
    json!([fixture_peer(), {
        "Identifier": "peer2pubkey==",
        "InterfaceIdentifier": "wg0",
        "PrivateKey": "k",
        "DisplayName": "Second Peer"
    }])
}

#[test]
fn get_by_id() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", format!("/api/v1/peer/by-id/{}", PEER_ID).as_str())
        .with_body(fixture_peer().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.peers().get_by_id(PEER_ID).unwrap();
    assert_eq!(result.identifier, PEER_ID);
    assert_eq!(result.interface_identifier, "wg0");
    mock.assert();
}

#[test]
fn get_by_interface() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/peer/by-interface/wg0")
        .with_body(fixture_peer_list().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.peers().get_by_interface("wg0").unwrap();
    assert_eq!(result.len(), 2);
    mock.assert();
}

#[test]
fn get_by_user() {
    let (mut server, client) = setup();
    let mock = server
        .mock("GET", "/api/v1/peer/by-user/uid-1234567")
        .with_body(fixture_peer_list().to_string())
        .with_header("content-type", "application/json")
        .create();

    let result = client.peers().get_by_user("uid-1234567").unwrap();
    assert_eq!(result.len(), 2);
    mock.assert();
}

#[test]
fn create_peer() {
    let (mut server, client) = setup();
    let mock = server
        .mock("POST", "/api/v1/peer/new")
        .with_body(fixture_peer().to_string())
        .with_header("content-type", "application/json")
        .create();

    let data: WgPeer = serde_json::from_value(fixture_peer()).unwrap();
    let result = client.peers().create(&data).unwrap();
    assert_eq!(result.identifier, PEER_ID);
    mock.assert();
}

#[test]
fn delete_peer() {
    let (mut server, client) = setup();
    let mock = server
        .mock("DELETE", format!("/api/v1/peer/by-id/{}", PEER_ID).as_str())
        .with_status(204)
        .create();

    client.peers().delete(PEER_ID).unwrap();
    mock.assert();
}
